### PR TITLE
Migrate `PyTuple` / `PyList` constructors

### PR DESCRIPTION
**Pull request by Icxolu**
Followup to #4559 

`PyList::new` and `PyTuple::new` become fallible.

It's worth double checking the `bad_clone_mem_leaks` tests, to make sure I didn't accidentally change the semantics there.

### CODE REVIEW

**Review by davidhewitt — COMMENTED** [review]
Thanks, it seems like the correct conclusion to me that the constructors have to become fallible, even it's a small breaking change.

We should mention this explicitly in its own newsfragment, I think.

**Comment by Icxolu** [comment]
Makes sense 👍

File: src/types/list.rs [hunk 1]
```diff
@@ -1054,9 +1059,13 @@ mod tests {
...
-        impl ToPyObject for Bad {
-            fn to_object(&self, py: Python<'_>) -> Py<PyAny> {
-                self.to_owned().into_py(py)
+        impl<'py> IntoPyObject<'py> for Bad {
+            type Target = PyInt;
+            type Output = crate::Bound<'py, Self::Target>;
+            type Error = Infallible;
+
+            fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
+                self.clone().0.into_pyobject(py)
```
Comment (davidhewitt) [inline src/types/list.rs:1068]: I wonder whether it's simpler to remove `Clone` from the picture and instead just have the panic produced in here?

IIRC, the main point of the test is to check that panics while constructing the list free stuff properly, `Clone` was probably just the bug report which came in or similar.

File: src/types/list.rs [hunk 2]
```diff
@@ -29,14 +29,14 @@ pub(crate) fn new_from_iter<'py>(
...
-    try_new_from_iter(py, &mut elements.map(Ok)).unwrap()
+    try_new_from_iter(py, &mut elements.map(|e| e.into_bound(py)).map(Ok)).unwrap()
 }
 
 #[inline]
 #[track_caller]
 pub(crate) fn try_new_from_iter<'py>(
     py: Python<'py>,
-    elements: &mut dyn ExactSizeIterator<Item = PyResult<PyObject>>,
+    elements: &mut dyn ExactSizeIterator<Item = PyResult<Bound<'py, PyAny>>>,
```
Comment (davidhewitt) [inline src/types/list.rs:39]: Do you think it's worth making these have a concrete iterator? I forget why we had to change it for sets 🤔

**Review by davidhewitt — APPROVED** [review]
Thanks, looks perfect! 🚀

File: src/types/list.rs (hunk 1 above)
Comment (Icxolu) [reply src/types/list.rs]: Yes, I think that makes it a bit clearer. I changed this, removed the now unnecessary `cfg` and renamed the test (also for `PyTuple`)

File: src/types/list.rs (hunk 2 above)
Comment (Icxolu) [reply src/types/list.rs:39]: Do you mean `impl` instead of `dyn` here? Or what concrete type are you referring to?

Comment (davidhewitt) [reply src/types/list.rs:39]: Yes, `impl Iterator`

**Review by davidhewitt — APPROVED** [review]
Super, thanks 👍 

I'll seek to get to the other reviews waiting on me either tonight or tomorrow evening

**Pull request merged**